third_party/move/move-compiler-v2/src/pipeline/reference_safety_processor_v2.rs

```diff
@@ -0,0 +1,1883 @@
+    parent: &'env LifeTimeAnalysis<'env>,
+    /// The code offset
+    code_offset: CodeOffset,
+    /// The attribute id at the code offset
+    attr_id: AttrId,
+    /// Lifetime information at the given code offset
+    alive: &'env LiveVarInfoAtCodeOffset,
+    /// Reference to the mutated analysis state
+    state: &'state mut LifetimeState,
+}
+
+/// Used to distinguish how a local is read
+#[derive(Clone, Copy, PartialEq, Eq)]
+enum ReadMode {
+    /// The local is moved
+    Move,
+    /// The local is copied
+    Copy,
+    /// The local is transferred as an argument to another function
+    Argument,
+}
+
+impl<'env> LifeTimeAnalysis<'env> {
+    fn new_step<'a>(
+        &'a self,
+        code_offset: CodeOffset,
+        attr_id: AttrId,
+        state: &'a mut LifetimeState,
+    ) -> LifetimeAnalysisStep {
+        let alive = self
+            .live_var_annotation
+            .get_live_var_info_at(code_offset)
+            .expect("live var info");
+        LifetimeAnalysisStep {
+            parent: self,
+            code_offset,
+            attr_id,
+            alive,
+            state,
+        }
+    }
+}
+
+// -------------------------------------------------------------------------------------------------
+// Analysing and Diagnosing
+
+impl<'env, 'state> LifetimeAnalysisStep<'env, 'state> {
+    /// Get the location associated with bytecode attribute.
+    fn loc(&self, id: AttrId) -> Loc {
+        self.target().get_bytecode_loc(id)
+    }
+
+    /// Returns the location of the current instruction
+    fn cur_loc(&self) -> Loc {
+        self.loc(self.attr_id)
+    }
+
+    /// Gets a string for a local to be displayed in error messages
+    fn display(&self, local: TempIndex) -> String {
+        self.target().get_local_name_for_error_message(local)
+    }
+
+    /// Display a non-empty set of temps. This finds the first printable representative, if any
+    fn display_set(&self, set: &BTreeSet<TempIndex>) -> String {
+        if let Some(temp) = set
+            .iter()
+            .find(|t| self.target().get_local_name_opt(**t).is_some())
+        {
+            self.display(*temp)
+        } else {
+            self.display(*set.first().expect("non empty"))
+        }
+    }
+
+    /// Returns "<prefix>`<name>` " if local has name, otherwise empty.
+    fn display_name_or_empty(&self, prefix: &str, local: TempIndex) -> String {
+        self.target()
+            .get_local_name_opt(local)
+            .map(|s| format!("{}`{}`", prefix, s))
+            .unwrap_or_default()
+    }
+
+    /// Get the type associated with local.
+    fn ty(&self, local: TempIndex) -> &Type {
+        self.target().get_local_type(local)
+    }
+
+    /// Returns true if the local is a reference.
+    fn is_ref(&self, local: TempIndex) -> bool {
+        self.ty(local).is_reference()
+    }
+
+    /// Check validness of reading a local. The read is not allowed if the local is borrowed. Returns true if valid.
```

Comment:
"the read is not allowed if the local is borrowed" => this seems incorrect without further qualifications?

Reply:
Outdated -- removed as the code is well documented